### PR TITLE
Allow station coordinates, in station and shape files, to be checked and compared

### LINKED ISSUE #6373: Checking Station coordinates
@shadrackkibet possibly to continue, or hand to a colleague.  There is an interesting investigation to do here.
There are 2 further steps with this dialogue and perhaps one step backwards.  

It currently does a TRUE or FALSE, but this actually needs more information than is currently provided.  That output should become optional.

a) Report the name of the country, or county or sub-county, etc.  (Not yet comparing with the station information.)  So, add a single receiver to the map information  - as well as the polygon.  This is usually the name of the county or country, etc.
Then after running, it knows the polygon so it reports the corresponding name.  This is a new variable in the station data frame.

b) Comparing with the station name.  This needs a new single receiver in the station part of the dialogue.  For example, with the current Kenya data this could be the variable with the county names.  Then there can be a string compare with the names found from the polygons.

There are then 3 possibilities.
a) The names match
b) The names are close, e.g. I think Homa Bay in the polygon file is Homabay in the station data.
c) The names don't match

I suggest the stringdist package is useful here.  It is implemented in the Prepare > Column: String > Distance dialogue!  Two columns can be compared.  I tried with a messed-up copy of the Station names and the cosine method.  I found:

a) names equal: cosine = 0
b) Homabay compared with Kisumu = 0.882
c) Compared with Homa bay = 0.050
d) Compared with Homa Bay = 0.157
e) Compared with Homeboy = 0.333 

Back to you to consider what could usefully be provided here!  I suggest some progress can be made now, and there could be more discussion too..

### PR DESCRIPTION
**Pull request by HawardKetoyoMsatsi**
@rdstern @africanmathsinitiative/developers This is Ready for review
The gadm36_KEN_shp and western_kenya_station_info datasets were used for testing however extra columns for Country, County, and Sub_County should be added to the station file for better testing.

### CODE REVIEW

**Comment by lloyddewit** [comment]
@N-thony please could you review? thanks

**Comment by N-thony** [comment]
> @N-thony please could you review? thanks

@lloyddewit yes, I'm happy to review.

File: instat/dlgLocatingPointsInShapeFile.vb [hunk 1]
```diff
@@ -135,4 +198,17 @@ Public Class dlgLocatingPointsInShapeFile
     Private Sub CoreControlsContentsChanged_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrReceiverLatitude.ControlContentsChanged, ucrReceiverLongitude.ControlContentsChanged, ucrSaveNewColumnName.ControlContentsChanged, ucrReceiverGeometry.ControlContentsChanged
```
Comment (N-thony) [inline instat/dlgLocatingPointsInShapeFile.vb:198]: add ucrReceiverStationFilePolygon and ucrReceiverShapeFilePolygon?

**Review by rdstern — CHANGES_REQUESTED** [review]
@haward I have trivial comments avout the dialogue.  Please can you check on the spacing of some of the controls.  Then we should also reflect on the polygon label - perhaps it should be Name or Name of Polygon.?
But the main problem - I wonder if you checked this is that it copies the wrong thing into the station file.  I used the Kenya data where there are 51 stations.
a) When I use the zero level it just copies Kenya over and leaves the rest as NA.
b) When I use level 1 - county level.  It copies over the 47 counties in alphabetical order. The last 4 are NA.  It should be copying the 51 county names that correspond to the location of each of the stations.

Once you get that right, i will check the other options.  The last option is a fuzzy comparison to cope with small changes in the names, between the station file and the geometry file.

**Comment by HawardKetoyoMsatsi** [comment]
@rdstern I have fixed the issue with level zero and Level one gadm files. Kindly have a look.

**Comment by rdstern** [comment]
@HawardKetoyoMsatsi this is great - it works.
There are some trivial points and one larger one.  
a) Change `Name of Polygon (Optional):` to `Location (Optional): `
b) Add tooltips, for the Station it could say "Name of the Country, County or Ward, if available in the station file"  For the polygon it could say "Name of the Country, County or Ward in the shapefile"
c) the names generated will change to say location rather than polygon.

Now a bigger point.  But I am inclined to merge this first, after you have done this.  An AIMS student is looking at mapping, including this problem, for his essay.  But don't close #6373 yet, because there is something there not done yet.  Unless you are desperate to do it, I would like the AIMS student to look at this problem and then advise you what to do.

**Comment by HawardKetoyoMsatsi** [comment]
@rdstern this is fixed

File: instat/dlgLocatingPointsInShapeFile.vb [hunk 2]
```diff
@@ -23,6 +23,11 @@ Public Class dlgLocatingPointsInShapeFile
     Private clsStCombineFunction As New RFunction
     Private clsConcFunction As New RFunction
     Private clsSubsetOperator As New ROperator
+    Private clsStAsIntegerFunction As New RFunction
+    Private clsExtractColumnStIntersectsFunction As New RFunction
+    Private clsDollarOperator As New ROperator
+    Private clsOpeningSubsetOperator, clsIsEqualToOperator, clsCompareColumnsdollarOperator,
```
Comment (lloyddewit) [inline instat/dlgLocatingPointsInShapeFile.vb:29]: `clsCompareColumnsdollarOperator` is not used, remove?

File: instat/dlgLocatingPointsInShapeFile.vb [hunk 3]
```diff
@@ -58,24 +66,43 @@ Public Class dlgLocatingPointsInShapeFile
...
+        ucrReceiverStationFilePolygon.Selector = ucrSelectorStationFile
+        ucrReceiverStationFilePolygon.SetParameterIsRFunction()
+
         ucrChkOmitMissing.SetParameter(New RParameter("na.fail", 3))
         ucrChkOmitMissing.SetText("Omit Missing Values")
         ucrChkOmitMissing.SetValuesCheckedAndUnchecked("FALSE", "TRUE")
         ucrChkOmitMissing.SetRDefault("TRUE")
 
+        ttStationFile.SetToolTip(ucrReceiverStationFilePolygon.txtReceiverSingle, "Name of the Country, County or Ward, if available in the station file")
+        ttStationFile.SetToolTip(ucrReceiverShapeFilePolygon.txtReceiverSingle, "Name of the Country, County or Ward in the shapefile")
```
Comment (lloyddewit) [inline instat/dlgLocatingPointsInShapeFile.vb:83]: 'Country, County or Ward` should not be capitalized
```suggestion
        ttStationFile.SetToolTip(ucrReceiverStationFilePolygon.txtReceiverSingle, "Name of the country, county or ward, if available in the station file")
        ttStationFile.SetToolTip(ucrReceiverShapeFilePolygon.txtReceiverSingle, "Name of the country, county or ward in the shapefile")
```

File: instat/dlgLocatingPointsInShapeFile.vb [hunk 4]
```diff
@@ -132,7 +200,17 @@ Public Class dlgLocatingPointsInShapeFile
         TestOKEnabled()
     End Sub
 
-    Private Sub CoreControlsContentsChanged_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrReceiverLatitude.ControlContentsChanged, ucrReceiverLongitude.ControlContentsChanged, ucrSaveNewColumnName.ControlContentsChanged, ucrReceiverGeometry.ControlContentsChanged
+    Private Sub CoreControlsContentsChanged_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrReceiverLatitude.ControlContentsChanged, ucrReceiverLongitude.ControlContentsChanged, ucrSaveNewColumnName.ControlContentsChanged, ucrReceiverGeometry.ControlContentsChanged, ucrReceiverShapeFilePolygon.ControlContentsChanged, ucrReceiverStationFilePolygon.ControlContentsChanged
```
Comment (lloyddewit) [inline instat/dlgLocatingPointsInShapeFile.vb:203]: Lines should not normally exceed 100 columns, add line breaks?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
Thanks, just a few small comments

**Comment by HawardKetoyoMsatsi** [comment]
@lloyddewit the above is fixe, please have a look.

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern There are just small changes since your last review, these should not have affected the functionality. Please could you test/approve? thanks

**Review by rdstern — APPROVED** [review]

**Pull request merged**